xvec/accessor.py

```diff
@@ -1258,6 +1258,75 @@ def extract_points(
                 )
         return result
 
+    def encode_cf(self) -> xr.Dataset:
+        """Encode geometry variables and associated CRS with CF conventions"""
+        import cf_xarray as cfxr
+
+        ds = self._obj.copy()
+
+        coords = self.geom_coords_indexed
+
+        # TODO: this could use geoxarray, but is quite simple in any case
+        # Adapted from rioxarray
+        # preserve ordering for roundtripping
+        unique_crs = []
+        for _, xi in sorted(coords.xindexes.items()):
+            if xi.crs not in unique_crs:
+                unique_crs.append(xi.crs)
+        if len(unique_crs) == 1:
+            grid_mappings = {unique_crs.pop(): "spatial_ref"}
+        else:
+            grid_mappings = {
+                crs_: f"spatial_ref_{i}" for i, crs_ in enumerate(unique_crs)
+            }
+
+        for crs, grid_mapping in grid_mappings.items():
+            grid_mapping_attrs = crs.to_cf()
+            # TODO: not all CRS can be represented by CF grid_mappings
+            # For now, we allow this.
+            # if "grid_mapping_name" not in grid_mapping_attrs:
+            #     raise ValueError
+            wkt_str = crs.to_wkt()
+            grid_mapping_attrs["spatial_ref"] = wkt_str
+            grid_mapping_attrs["crs_wkt"] = wkt_str
+            ds.coords[grid_mapping] = xr.Variable(
+                dims=(), data=0, attrs=grid_mapping_attrs
+            )
+
+        for name, coord in coords.items():
+            dims = set(coord.dims)
+            index = coords.xindexes[name]
+            varnames = (k for k, v in ds._variables.items() if dims & set(v.dims))
+            for name in varnames:
+                ds._variables[name].attrs["grid_mapping"] = grid_mappings[index.crs]
+        encoded = cfxr.geometry.encode_geometries(ds)
+        return encoded
+
+    def decode_cf(self) -> xr.Dataset:
+        import cf_xarray as cfxr
+
+        decoded = cfxr.geometry.decode_geometries(self._obj.copy())
+        crs = {
+            name: CRS.from_cf(var.attrs)
+            for name, var in decoded._variables.items()
+            if "crs_wkt" in var.attrs or "grid_mapping_name" in var.attrs
+        }
+        dims = decoded.xvec.geom_coords.dims
+        for dim in dims:
+            decoded = (
+                decoded.set_xindex(dim) if dim not in decoded._indexes else decoded
+            )
+            decoded = decoded.xvec.set_geom_indexes(
+                dim, crs=crs.get(decoded[dim].attrs.get("grid_mapping", None))
+            )
```

Comment:
This is the key line. it always sets the index, we do not record which geometry dims were indexed at encode-time. As an aside it'd be nice for `set_geom_indexes` to understand the `grid_mapping` convention. WDYT?

Reply:
>  it always sets the index, we do not record which geometry dims were indexed at encode-time. What should we do here?

Is that an issue if we just index all geom dims encoded in the file?

 > As an aside it'd be nice for set_geom_indexes to understand the grid_mapping convention. WDYT?

Not against but I don't really know what would it mean implementation-wise. Maybe just a simple call to [`pyproj.CRS.from_cf`](https://pyproj4.github.io/pyproj/dev/build_crs_cf.html#build-the-crs-from-cf-grid-mapping)?

> set_geom_indexes learns how to interpret the grid_mapping convention so CRS is set properly by default.

That would be preferable. Not a fan of asking users to set indexes after reading what already was indexed before writing.